File: app/src/main/java/forpdateam/ru/forpda/fragments/theme/editpost/PollChoicesAdapter.java
package forpdateam.ru.forpda.fragments.theme.editpost;

import android.content.DialogInterface;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.api.theme.editpost.models.EditPoll;
import forpdateam.ru.forpda.utils.SimpleTextWatcher;

/**
 * Created by radiationx on 28.07.17.
 */


public class PollChoicesAdapter extends RecyclerView.Adapter<PollChoicesAdapter.ViewHolder> {
    private List<EditPoll.Choice> choices = new ArrayList<>();
    private EditPoll poll;
    private EditPoll.Question question;

    public PollChoicesAdapter(EditPoll.Question question, EditPoll poll) {
        this.choices = question.getChoices();
        this.poll = poll;
        this.question = question;
    }

    public PollChoicesAdapter() {
    }

    public void add(EditPoll.Choice choice) {
        if (this.choices.size() < poll.getMaxChoices()) {
            question.increaseIndexOffset();
            choice.setIndex(question.getIndexOffset() + question.getBaseIndexOffset());
            this.choices.add(choice);
            //notifyItemInserted(choices.indexOf(choice));
            notifyDataSetChanged();
        } else {
            Toast.makeText(App.getContext(), "Максимальное кол-во ответов: " + poll.getMaxChoices(), Toast.LENGTH_SHORT).show();
        }
    }

    public EditPoll.Choice getItem(int position) {
        return choices.get(position);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.edit_poll_choice, parent, false);
        return new ViewHolder(v, new MyCustomEditTextListener());
    }

    @Override
    public int getItemCount() {
        return choices.size();
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        EditPoll.Choice item = getItem(holder.getAdapterPosition());
        assert item != null;

        holder.myCustomEditTextListener.updatePosition(holder.getAdapterPosition());
        holder.title.getEditText().setText(item.getTitle());
        holder.title.setHint("Ответ " + (holder.getAdapterPosition() + 1));


    }


    public class ViewHolder extends RecyclerView.ViewHolder {
        public TextInputLayout title;
        public ImageButton delete;
        public MyCustomEditTextListener myCustomEditTextListener;

        public ViewHolder(View v, MyCustomEditTextListener myCustomEditTextListener) {
            super(v);
            title = (TextInputLayout) v.findViewById(R.id.poll_choice_title);
            delete = (ImageButton) v.findViewById(R.id.poll_choice_delete);

            this.myCustomEditTextListener = myCustomEditTextListener;
            this.title.getEditText().addTextChangedListener(myCustomEditTextListener);
            delete.setOnClickListener(v1 -> {
                new AlertDialog.Builder(v.getContext())
                        .setMessage("Удалить ответ?")
                        .setPositiveButton("Да", (dialog, which) -> {
                            EditPoll.Choice choice = choices.get(getLayoutPosition());
                            //notifyItemRemoved(getLayoutPosition());
                            if (choice.getIndex() > question.getBaseIndexOffset()) {
                                int start = choice.getIndex();
                                int end = question.getBaseIndexOffset() + question.getIndexOffset();
                                for (int i = start; i <= end; i++) {
                                    EditPoll.Choice c = EditPoll.findChoiceByIndex(question, i);
                                    if (c != null) {
                                        c.setIndex(c.getIndex() - 1);
                                    }
                                }
                                question.reduceIndexOffset();
                            }
                            choices.remove(getLayoutPosition());
                            notifyDataSetChanged();
                        })
                        .setNegativeButton("Нет", null)
                        .show();

            });
        }
    }

    private class MyCustomEditTextListener extends SimpleTextWatcher {
        private int position;

        public void updatePosition(int position) {
            this.position = position;
        }

        @Override
        public void onTextChanged(CharSequence charSequence, int i, int i2, int i3) {
            choices.get(position).setTitle(charSequence.toString());
        }
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/editpost/PollChoicesAdapter.kt
package forpdateam.ru.forpda.ui.fragments.editpost

import android.content.Context
import android.content.DialogInterface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.RecyclerView
import by.kirich1409.viewbindingdelegate.viewBinding
import forpdateam.ru.forpda.App.Companion.get
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.common.simple.SimpleTextWatcher
import forpdateam.ru.forpda.databinding.EditPollChoiceBinding
import forpdateam.ru.forpda.entity.remote.editpost.EditPoll
import forpdateam.ru.forpda.entity.remote.editpost.EditPoll.Companion.findChoiceByIndex
import forpdateam.ru.forpda.entity.remote.editpost.EditPoll.Question

/**
 * Created by radiationx on 28.07.17.
 */
class PollChoicesAdapter : RecyclerView.Adapter<PollChoicesAdapter.ViewHolder> {
    private var choices: ArrayList<EditPoll.Choice> = ArrayList<EditPoll.Choice>()
    private var poll: EditPoll? = null
    private var question: Question? = null

    constructor(question: Question, poll: EditPoll?) {
        this.choices = question.getChoices()
        this.poll = poll
        this.question = question
    }

    fun add(context: Context, choice: EditPoll.Choice) {
        if (choices.size < poll!!.maxChoices) {
            question!!.increaseIndexOffset()
            choice.index = question!!.indexOffset + question!!.baseIndexOffset
            choices.add(choice)
            //notifyItemInserted(choices.indexOf(choice));
            notifyDataSetChanged()
        } else {
            Toast.makeText(
                context,
                String.format(get().getString(R.string.poll_answers_Max), poll!!.maxChoices),
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    fun getItem(position: Int): EditPoll.Choice {
        return choices[position]
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v =
            LayoutInflater.from(parent.context).inflate(R.layout.edit_poll_choice, parent, false)
        return ViewHolder(v, MyCustomEditTextListener())
    }

    override fun getItemCount(): Int {
        return choices.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(
        v: View,
        private val myCustomEditTextListener: MyCustomEditTextListener
    ) : RecyclerView.ViewHolder(v) {

        private val binding by viewBinding<EditPollChoiceBinding>()

        init {
            binding.pollChoiceTitle.editText!!.addTextChangedListener(myCustomEditTextListener)
            binding.pollChoiceDelete.setOnClickListener { v1: View? ->
                AlertDialog.Builder(v.context)
                    .setMessage(R.string.ask_delete_answer)
                    .setPositiveButton(R.string.ok) { dialog: DialogInterface?, which: Int ->
                        val choice = choices[layoutPosition]
                        //notifyItemRemoved(getLayoutPosition());
                        if (choice.index > question!!.baseIndexOffset) {
                            val start = choice.index
                            val end = question!!.baseIndexOffset + question!!.indexOffset
                            for (i in start..end) {
                                val c = findChoiceByIndex(
                                    question!!, i
                                )
                                if (c != null) {
                                    c.index = c.index - 1
                                }
                            }
                            question!!.reduceIndexOffset()
                        }
                        choices.removeAt(layoutPosition)
                        notifyDataSetChanged()
                    }
                    .setNegativeButton(R.string.no, null)
                    .show()
            }
        }

        fun bind(item: EditPoll.Choice) {
            myCustomEditTextListener.updatePosition(adapterPosition)
            binding.pollChoiceTitle.editText!!.setText(item.title)
            binding.pollChoiceTitle.hint = String.format(
                get().getString(R.string.poll_answer_Pos),
                adapterPosition + 1
            )
        }
    }

    inner class MyCustomEditTextListener : SimpleTextWatcher() {
        private var position = 0

        fun updatePosition(position: Int) {
            this.position = position
        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            choices[position].title = s.toString()
        }
    }
}
